x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,367 @@
+        });
+
+        let param = Param::from(value);
+        let expected_param = Param::U64(value);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
+
+    #[test]
+    fn convert_string_param() {
+        let value = String::from("hello world");
+        let expected_param_type = "string";
+
+        let expected_json = json!({
+            "type": expected_param_type,
+            "value": &b64.encode(borsh::to_vec(&value).unwrap()),
+        });
+
+        let param = Param::from(value.clone());
+        let expected_param = Param::String(value);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
+
+    #[test]
+    fn convert_id_param() {
+        let value: usize = 42;
+        let expected_param_type = "id";
+        let expected_value = value.to_le_bytes();
+
+        let expected_json = json!({
+            "type": expected_param_type,
+            "value": &b64.encode(expected_value)
+        });
+
+        let id = Id::from(value);
+        let param = Param::from(id);
+        let expected_param = Param::Id(id);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
+
+    #[test]
+    fn convert_bool_param() {
+        let value = false;
+        let expected_value = value as u8;
+
+        let expected_json = json!({
+            "type": "bool",
+            "value": &b64.encode(vec![expected_value]),
+        });
+
+        let param = Param::from(value);
+        let expected_param = Param::Bool(value);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
+
+    #[test]
+    fn convert_bytes_param() {
+        let value = vec![12, 34, 56, 78, 90];
+
+        let expected_json = json!({
+            "type": "bytes",
+            "value": &b64.encode(borsh::to_vec(&value).unwrap()),
+        });
+
+        let param = Param::from(value.clone());
+        let expected_param = Param::Bytes(value);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
+
+    #[test]
+    fn convert_address_param() {
+        let value = Address::default();
+
+        let expected_json = json!({
+            "type": "address",
+            "value": &b64.encode(value),
+        });
+
+        let param = Param::from(value);
+        let expected_param = Param::Address(value);
+
+        assert_eq!(param, expected_param);
+
+        let output_json = serde_json::to_value(&param).unwrap();
+
+        assert_eq!(output_json, expected_json);
+    }
```

Comment:
All these tests (should we decide to keep them) belong in `param.rs`, not here. 

However, I question their value. If we change the way we serialize, we have to change all the tests. The tests will only fail if we make an explicit change as it stands. All of the logic lives in `serde_derive`, we're only testing how we tag our structs. 

I think this is more of a maintenance burden than adding any value. Tests like these only really make sense when we are making promises about the crate-API. Right now it's just a maintenance burden.